File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/player_screen/PlayerPresenter.java
package com.github.anrimian.musicplayer.ui.player_screen;

import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor;
import com.github.anrimian.musicplayer.domain.interactors.player.PlayerScreenInteractor;
import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueEvent;
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem;
import com.github.anrimian.musicplayer.domain.models.player.PlayerState;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.domain.utils.ListUtils;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import moxy.MvpPresenter;

import static com.github.anrimian.musicplayer.domain.models.utils.CompositionHelper.areSourcesTheSame;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.mapList;

/**
 * Created on 02.11.2017.
 */

public class PlayerPresenter extends MvpPresenter<PlayerView> {

    private final LibraryPlayerInteractor playerInteractor;
    private final PlayListsInteractor playListsInteractor;
    private final PlayerScreenInteractor playerScreenInteractor;
    private final ErrorParser errorParser;
    private final Scheduler uiScheduler;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();
    private final CompositeDisposable batterySafeDisposable = new CompositeDisposable();

    private List<PlayQueueItem> playQueue = new ArrayList<>();

    private PlayQueueItem currentItem;
    private int currentPosition = -1;
    private boolean isDragging;

    private boolean isCoversEnabled = false;

    private final List<Composition> compositionsForPlayList = new LinkedList<>();
    private final List<Composition> compositionsToDelete = new LinkedList<>();

    private short numberOfQueueUpdatesToIgnore = 0;

    public PlayerPresenter(LibraryPlayerInteractor musicPlayerInteractor,
                           PlayListsInteractor playListsInteractor,
                           PlayerScreenInteractor playerScreenInteractor,
                           ErrorParser errorParser,
                           Scheduler uiScheduler) {
        this.playerInteractor = musicPlayerInteractor;
        this.playListsInteractor = playListsInteractor;
        this.playerScreenInteractor = playerScreenInteractor;
        this.errorParser = errorParser;
        this.uiScheduler = uiScheduler;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        if (playerScreenInteractor.isPlayerPanelOpen()) {
            getViewState().expandBottomPanel();
        } else {
            getViewState().collapseBottomPanel();
        }
        subscribeOnUiSettings();
        subscribeOnRandomMode();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onStart() {
        subscribeOnRepeatMode();
        subscribeOnPlayerStateChanges();
        subscribeOnPlayQueue();
        subscribeOnCurrentCompositionChanging();
        subscribeOnCurrentPosition();
        subscribeOnTrackPositionChanging();
    }

    void onStop() {
        batterySafeDisposable.clear();
    }

    void onCurrentScreenRequested() {
        getViewState().showDrawerScreen(playerScreenInteractor.getSelectedDrawerScreen(),
                playerScreenInteractor.getSelectedPlayListScreenId());
    }

    void onOpenPlayQueueClicked() {
        playerScreenInteractor.setPlayerPanelOpen(true);
    }

    void onBottomPanelExpanded() {
        playerScreenInteractor.setPlayerPanelOpen(true);
        getViewState().expandBottomPanel();
    }

    void onBottomPanelCollapsed() {
        playerScreenInteractor.setPlayerPanelOpen(false);
        getViewState().collapseBottomPanel();
    }

    void onDrawerScreenSelected(int screenId) {
        playerScreenInteractor.setSelectedDrawerScreen(screenId);
        getViewState().showDrawerScreen(screenId, 0);
    }

    void onLibraryScreenSelected() {
        getViewState().showLibraryScreen(playerScreenInteractor.getSelectedLibraryScreen());
    }

    void onPlayButtonClicked() {
        playerInteractor.play();
    }

    void onStopButtonClicked() {
        playerInteractor.pause();
    }

    void onSkipToPreviousButtonClicked() {
        playerInteractor.skipToPrevious();
    }

    void onSkipToNextButtonClicked() {
        playerInteractor.skipToNext();
    }

    void onRepeatModeChanged(int mode) {
        playerInteractor.setRepeatMode(mode);
    }

    void onRandomPlayingButtonClicked(boolean enable) {
        playerInteractor.setRandomPlayingEnabled(enable);
    }

    void onShareCompositionButtonClicked() {
        getViewState().showShareMusicDialog(currentItem.getComposition());
    }

    void onCompositionItemClicked(int position, PlayQueueItem item) {
        this.currentPosition = position;
        this.currentItem = item;
        playerInteractor.skipToItem(item);

        onCurrentCompositionChanged(item, 0);
    }

    void onQueueItemIconClicked(int position, PlayQueueItem playQueueItem) {
        if (playQueueItem.equals(currentItem)) {
            playerInteractor.playOrPause();
        } else {
            onCompositionItemClicked(position, playQueueItem);
            playerInteractor.play();
        }
    }

    void onTrackRewoundTo(int progress) {
        playerInteractor.seekTo(progress);
    }

    void onDeleteCompositionButtonClicked(Composition composition) {
        compositionsToDelete.clear();
        compositionsToDelete.add(composition);
        getViewState().showConfirmDeleteDialog(compositionsToDelete);
    }

    void onDeleteCurrentCompositionButtonClicked() {
        compositionsToDelete.clear();
        compositionsToDelete.add(currentItem.getComposition());
        getViewState().showConfirmDeleteDialog(compositionsToDelete);
    }

    void onAddQueueItemToPlayListButtonClicked(Composition composition) {
        compositionsForPlayList.clear();
        compositionsForPlayList.add(composition);
        getViewState().showSelectPlayListDialog();
    }

    void onAddCurrentCompositionToPlayListButtonClicked() {
        compositionsForPlayList.clear();
        compositionsForPlayList.add(currentItem.getComposition());
        getViewState().showSelectPlayListDialog();
    }

    void onPlayListForAddingSelected(PlayList playList) {
        addPreparedCompositionsToPlayList(playList);
    }

    void onPlayListForAddingCreated(PlayList playList) {
        List<Composition> compositionsToAdd = mapList(playQueue, PlayQueueItem::getComposition);
        playListsInteractor.addCompositionsToPlayList(compositionsToAdd, playList)
                .observeOn(uiScheduler)
                .subscribe(() -> getViewState().showAddingToPlayListComplete(playList, compositionsToAdd),
                        this::onAddingToPlayListError);
    }

    void onDeleteCompositionsDialogConfirmed() {
        deletePreparedCompositions();
    }

    void onSeekStart() {
        playerInteractor.onSeekStarted();
    }

    void onSeekStop(int progress) {
        playerInteractor.onSeekFinished(progress);
    }

    void onItemSwipedToDelete(Integer position) {
        deletePlayQueueItem(playQueue.get(position));
    }

    void onDeleteQueueItemClicked(PlayQueueItem item) {
        deletePlayQueueItem(item);
    }

    void onItemMoved(int from, int to) {
        if (from < to) {
            for (int i = from; i < to; i++) {
                swapItems(i, i + 1);
            }
        } else {
            for (int i = from; i > to; i--) {
                swapItems(i, i - 1);
            }
        }
    }

    void onEditCompositionButtonClicked() {
        getViewState().startEditCompositionScreen(currentItem.getComposition().getId());
    }

    void onRestoreDeletedItemClicked() {
        playerInteractor.restoreDeletedItem()
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
    }

    void onClearPlayQueueClicked() {
        playerInteractor.clearPlayQueue();
    }

    void onFastSeekForwardCalled() {
        playerInteractor.fastSeekForward();
    }

    void onFastSeekBackwardCalled() {
        playerInteractor.fastSeekBackward();
    }

    void onDragStarted(int position) {
        isDragging = true;
    }

    void onDragEnded(int position) {
        isDragging = false;
    }

    private void swapItems(int from, int to) {
        if (!ListUtils.isIndexInRange(playQueue, from) || !ListUtils.isIndexInRange(playQueue, to)) {
            return;
        }
        PlayQueueItem fromItem = playQueue.get(from);
        PlayQueueItem toItem = playQueue.get(to);

        Collections.swap(playQueue, from, to);
        getViewState().notifyItemMoved(from, to);

        numberOfQueueUpdatesToIgnore++;
        playerInteractor.swapItems(fromItem, toItem);
    }

    private void deletePlayQueueItem(PlayQueueItem item) {
        playerInteractor.removeQueueItem(item)
                .observeOn(uiScheduler)
                .subscribe(getViewState()::showDeletedItemMessage);
    }

    private void subscribeOnRepeatMode() {
        batterySafeDisposable.add(playerInteractor.getRepeatModeObservable()
                .observeOn(uiScheduler)
                .subscribe(getViewState()::showRepeatMode));
    }

    private void addPreparedCompositionsToPlayList(PlayList playList) {
        playListsInteractor.addCompositionsToPlayList(compositionsForPlayList, playList)
                .observeOn(uiScheduler)
                .subscribe(() -> onAddingToPlayListCompleted(playList),
                        this::onAddingToPlayListError);
    }

    private void deletePreparedCompositions() {
        playerInteractor.deleteCompositions(compositionsToDelete)
                .observeOn(uiScheduler)
                .subscribe(this::onDeleteCompositionsSuccess, this::onDeleteCompositionError);
    }

    private void onDeleteCompositionsSuccess() {
        getViewState().showDeleteCompositionMessage(compositionsToDelete);
        compositionsToDelete.clear();
    }

    private void onDeleteCompositionError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showDeleteCompositionError(errorCommand);
        compositionsToDelete.clear();
    }

    private void onAddingToPlayListError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showAddingToPlayListError(errorCommand);
    }

    private void onAddingToPlayListCompleted(PlayList playList) {
        getViewState().showAddingToPlayListComplete(playList, compositionsForPlayList);
        compositionsForPlayList.clear();
    }

    private void subscribeOnCurrentCompositionChanging() {
        batterySafeDisposable.add(playerInteractor.getCurrentQueueItemObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onPlayQueueEventReceived));
    }

    private void subscribeOnCurrentPosition() {
        batterySafeDisposable.add(playerInteractor.getCurrentItemPositionObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onItemPositionReceived));
    }

    private void onItemPositionReceived(int position) {
        if (!isDragging && currentPosition != position) {
            currentPosition = position;
            getViewState().scrollQueueToPosition(position);
        }
    }

    private void onPlayQueueEventReceived(PlayQueueEvent playQueueEvent) {
        PlayQueueItem newItem = playQueueEvent.getPlayQueueItem();
        if (currentItem == null
                || !currentItem.equals(newItem)
                || !areSourcesTheSame(newItem.getComposition(), currentItem.getComposition())) {
            onCurrentCompositionChanged(newItem, playQueueEvent.getTrackPosition());
        }
    }

    private void onCurrentCompositionChanged(PlayQueueItem newItem, long trackPosition) {
        getViewState().showCurrentQueueItem(newItem, isCoversEnabled);
        if (newItem != null
                && (!newItem.equals(currentItem) || newItem.getComposition().getDuration() != currentItem.getComposition().getDuration())) {
            getViewState().showTrackState(trackPosition, newItem.getComposition().getDuration());
        }

        this.currentItem = newItem;
    }

    private void subscribeOnPlayerStateChanges() {
        batterySafeDisposable.add(playerInteractor.getPlayerStateObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onPlayerStateChanged));
    }

    private void onPlayerStateChanged(PlayerState playerState) {
        switch (playerState) {
            case PLAY: {
                getViewState().showPlayState();
                return;
            }
            default: {
                getViewState().showStopState();
            }
        }
    }

    private void subscribeOnPlayQueue() {
        batterySafeDisposable.add(playerInteractor.getPlayQueueObservable()
                .observeOn(uiScheduler)
                .filter(o -> isPlayQueueEmitAllowed())
                .subscribe(this::onPlayQueueChanged, this::onPlayQueueReceivingError));
    }

    private boolean isPlayQueueEmitAllowed() {
        boolean isAllowed = numberOfQueueUpdatesToIgnore <= 0;
        if (numberOfQueueUpdatesToIgnore > 0) {
            numberOfQueueUpdatesToIgnore--;
        }
        return isAllowed;
    }

    private void onPlayQueueChanged(List<PlayQueueItem> list) {
        playQueue = list;
        getViewState().showPlayQueueSubtitle(playQueue.size());
        getViewState().setMusicControlsEnabled(!playQueue.isEmpty());
        getViewState().updatePlayQueue(list);
    }

    private void onPlayQueueReceivingError(Throwable throwable) {
        errorParser.parseError(throwable);
        getViewState().setMusicControlsEnabled(false);
    }

    private void subscribeOnTrackPositionChanging() {
        batterySafeDisposable.add(playerInteractor.getTrackPositionObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onTrackPositionChanged));
    }

    private void onTrackPositionChanged(Long currentPosition) {
        if (currentItem != null) {
            long duration = currentItem.getComposition().getDuration();
            getViewState().showTrackState(currentPosition, duration);
        }
    }

    private void subscribeOnUiSettings() {
        presenterDisposable.add(playerScreenInteractor.getCoversEnabledObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onUiSettingsReceived, errorParser::logError));
    }

    private void onUiSettingsReceived(boolean isCoversEnabled) {
        this.isCoversEnabled = isCoversEnabled;
        getViewState().setPlayQueueCoversEnabled(isCoversEnabled);
        if (currentItem != null) {
            getViewState().showCurrentQueueItem(currentItem, isCoversEnabled);
        }
    }

    private void onDefaultError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showErrorMessage(errorCommand);
    }

    private void subscribeOnRandomMode() {
        presenterDisposable.add(playerInteractor.getRandomPlayingObservable()
                .observeOn(uiScheduler)
                .subscribe(getViewState()::showRandomPlayingButton));
    }
}

File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/player_screen/PlayerPresenter.kt
package com.github.anrimian.musicplayer.ui.player_screen

import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor
import com.github.anrimian.musicplayer.domain.interactors.player.PlayerScreenInteractor
import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueEvent
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem
import com.github.anrimian.musicplayer.domain.models.player.PlayerState
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.domain.models.utils.CompositionHelper
import com.github.anrimian.musicplayer.domain.utils.ListUtils
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.ListDragFilter
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.disposables.CompositeDisposable
import java.util.*

/**
 * Created on 02.11.2017.
 */
class PlayerPresenter(
        private val playerInteractor: LibraryPlayerInteractor,
        private val playListsInteractor: PlayListsInteractor,
        private val playerScreenInteractor: PlayerScreenInteractor,
        errorParser: ErrorParser,
        uiScheduler: Scheduler
) : AppPresenter<PlayerView>(uiScheduler, errorParser) {
    
    private val batterySafeDisposable = CompositeDisposable()

    private val listDragFilter = ListDragFilter()

    private var playQueue: List<PlayQueueItem> = ArrayList()
    private var currentItem: PlayQueueItem? = null
    private var currentPosition = -1
    
    private var isDragging = false
    private var isCoversEnabled = false
    
    private val compositionsForPlayList = LinkedList<Composition>()

    private var lastDeleteAction: Completable? = null
    
    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.setButtonPanelState(playerScreenInteractor.isPlayerPanelOpen)
        subscribeOnUiSettings()
        subscribeOnRandomMode()
        subscribeOnSpeedAvailableState()
        subscribeOnSpeedState()
    }

    fun onStart() {
        subscribeOnRepeatMode()
        subscribeOnPlayerStateChanges()
        subscribeOnPlayQueue()
        subscribeOnCurrentCompositionChanging()
        subscribeOnCurrentPosition()
        subscribeOnTrackPositionChanging()
        subscribeOnSleepTimerTime()
    }

    fun onStop() {
        batterySafeDisposable.clear()
    }

    fun onCurrentScreenRequested() {
        viewState.showDrawerScreen(playerScreenInteractor.selectedDrawerScreen,
                playerScreenInteractor.selectedPlayListScreenId)
    }

    fun onOpenPlayQueueClicked() {
        playerScreenInteractor.isPlayerPanelOpen = true
    }

    fun onBottomPanelExpanded() {
        playerScreenInteractor.isPlayerPanelOpen = true
        viewState.setButtonPanelState(true)
    }

    fun onBottomPanelCollapsed() {
        playerScreenInteractor.isPlayerPanelOpen = false
        viewState.setButtonPanelState(false)
    }

    fun onDrawerScreenSelected(screenId: Int) {
        playerScreenInteractor.selectedDrawerScreen = screenId
        viewState.showDrawerScreen(screenId, 0)
    }

    fun onLibraryScreenSelected() {
        viewState.showLibraryScreen(playerScreenInteractor.selectedLibraryScreen)
    }

    fun onPlayButtonClicked() {
        playerInteractor.play()
    }

    fun onStopButtonClicked() {
        playerInteractor.pause()
    }

    fun onSkipToPreviousButtonClicked() {
        playerInteractor.skipToPrevious()
    }

    fun onSkipToNextButtonClicked() {
        playerInteractor.skipToNext()
    }

    fun onRepeatModeChanged(mode: Int) {
        playerInteractor.repeatMode = mode
    }

    fun onRandomPlayingButtonClicked(enable: Boolean) {
        playerInteractor.isRandomPlayingEnabled = enable
    }

    fun onShareCompositionButtonClicked() {
        viewState.showShareMusicDialog(currentItem!!.composition)
    }

    fun onQueueItemClicked(position: Int, item: PlayQueueItem) {
        if (item == currentItem) {
            playerInteractor.playOrPause()
            return
        }
        currentPosition = position
        currentItem = item
        playerInteractor.skipToItem(item)
        onCurrentCompositionChanged(item, 0)
    }

    fun onQueueItemIconClicked(position: Int, playQueueItem: PlayQueueItem) {
        if (playQueueItem == currentItem) {
            playerInteractor.playOrPause()
        } else {
            onQueueItemClicked(position, playQueueItem)
            playerInteractor.play()
        }
    }

    fun onTrackRewoundTo(progress: Int) {
        playerInteractor.seekTo(progress.toLong())
    }

    fun onDeleteCompositionButtonClicked(composition: Composition) {
        viewState.showConfirmDeleteDialog(listOf(composition))
    }

    fun onDeleteCurrentCompositionButtonClicked() {
        viewState.showConfirmDeleteDialog(listOf(currentItem!!.composition))
    }

    fun onAddQueueItemToPlayListButtonClicked(composition: Composition) {
        compositionsForPlayList.clear()
        compositionsForPlayList.add(composition)
        viewState.showSelectPlayListDialog()
    }

    fun onAddCurrentCompositionToPlayListButtonClicked() {
        compositionsForPlayList.clear()
        compositionsForPlayList.add(currentItem!!.composition)
        viewState.showSelectPlayListDialog()
    }

    fun onPlayListForAddingSelected(playList: PlayList) {
        addPreparedCompositionsToPlayList(playList)
    }

    fun onPlayListForAddingCreated(playList: PlayList?) {
        val compositionsToAdd = playQueue.map(PlayQueueItem::getComposition)
        playListsInteractor.addCompositionsToPlayList(compositionsToAdd, playList)
                .subscribeOnUi(
                        { viewState.showAddingToPlayListComplete(playList, compositionsToAdd) },
                        this::onAddingToPlayListError
                )
    }

    fun onDeleteCompositionsDialogConfirmed(compositionsToDelete: List<Composition>) {
        deletePreparedCompositions(compositionsToDelete)
    }

    fun onSeekStart() {
        playerInteractor.onSeekStarted()
    }

    fun onSeekStop(progress: Int) {
        playerInteractor.onSeekFinished(progress.toLong())
    }

    fun onItemSwipedToDelete(position: Int) {
        deletePlayQueueItem(playQueue[position])
    }

    fun onDeleteQueueItemClicked(item: PlayQueueItem?) {
        deletePlayQueueItem(item)
    }

    fun onItemMoved(from: Int, to: Int) {
        if (from < to) {
            for (i in from until to) {
                swapItems(i, i + 1)
            }
        } else {
            for (i in from downTo to + 1) {
                swapItems(i, i - 1)
            }
        }
    }

    fun onEditCompositionButtonClicked() {
        viewState.startEditCompositionScreen(currentItem!!.composition.id)
    }

    fun onRestoreDeletedItemClicked() {
        playerInteractor.restoreDeletedItem().justSubscribe(this::onDefaultError)
    }

    fun onClearPlayQueueClicked() {
        playerInteractor.clearPlayQueue()
    }

    fun onFastSeekForwardCalled() {
        playerInteractor.fastSeekForward()
    }

    fun onFastSeekBackwardCalled() {
        playerInteractor.fastSeekBackward()
    }

    fun onDragStarted() {
        isDragging = true
    }

    fun onDragEnded() {
        isDragging = false
    }

    fun onRetryFailedDeleteActionClicked() {
        if (lastDeleteAction != null) {
            lastDeleteAction!!
                    .doFinally { lastDeleteAction = null }
                    .subscribe({}, this::onDeleteCompositionError)
        }
    }

    fun onPlaybackSpeedSelected(speed: Float) {
        viewState.displayPlaybackSpeed(speed)
        playerInteractor.playbackSpeed = speed
    }

    private fun swapItems(from: Int, to: Int) {
        if (!ListUtils.isIndexInRange(playQueue, from) || !ListUtils.isIndexInRange(playQueue, to)) {
            return
        }
        val fromItem = playQueue[from]
        val toItem = playQueue[to]
        Collections.swap(playQueue, from, to)
        viewState.notifyItemMoved(from, to)

        listDragFilter.increaseEventsToSkip()
        playerInteractor.swapItems(fromItem, toItem)
    }

    private fun deletePlayQueueItem(item: PlayQueueItem?) {
        playerInteractor.removeQueueItem(item).unsafeSubscribeOnUi(viewState::showDeletedItemMessage)
    }

    private fun subscribeOnRepeatMode() {
        batterySafeDisposable.add(playerInteractor.repeatModeObservable
                .observeOn(uiScheduler)
                .subscribe(viewState::showRepeatMode))
    }

    private fun addPreparedCompositionsToPlayList(playList: PlayList) {
        playListsInteractor.addCompositionsToPlayList(compositionsForPlayList, playList)
                .subscribeOnUi({ onAddingToPlayListCompleted(playList) }, this::onAddingToPlayListError)
    }

    private fun deletePreparedCompositions(compositionsToDelete: List<Composition>) {
        lastDeleteAction = playerInteractor.deleteCompositions(compositionsToDelete)
                .observeOn(uiScheduler)
                .doOnComplete { onDeleteCompositionsSuccess(compositionsToDelete) }
        lastDeleteAction!!.justSubscribe(this::onDeleteCompositionError)
    }

    private fun onDeleteCompositionsSuccess(compositionsToDelete: List<Composition>) {
        viewState.showDeleteCompositionMessage(compositionsToDelete)
    }

    private fun onDeleteCompositionError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showDeleteCompositionError(errorCommand)
    }

    private fun onAddingToPlayListError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showAddingToPlayListError(errorCommand)
    }

    private fun onAddingToPlayListCompleted(playList: PlayList) {
        viewState.showAddingToPlayListComplete(playList, compositionsForPlayList)
        compositionsForPlayList.clear()
    }

    private fun subscribeOnCurrentCompositionChanging() {
        batterySafeDisposable.add(playerInteractor.currentQueueItemObservable
                .observeOn(uiScheduler)
                .subscribe(this::onPlayQueueEventReceived))
    }

    private fun subscribeOnCurrentPosition() {
        batterySafeDisposable.add(playerInteractor.currentItemPositionObservable
                .observeOn(uiScheduler)
                .subscribe(this::onItemPositionReceived))
    }

    private fun onItemPositionReceived(position: Int) {
        if (!isDragging && currentPosition != position) {
            currentPosition = position
            viewState.scrollQueueToPosition(position)
        }
    }

    private fun onPlayQueueEventReceived(playQueueEvent: PlayQueueEvent) {
        val newItem = playQueueEvent.playQueueItem
        if (currentItem == null || currentItem != newItem
                || !CompositionHelper.areSourcesTheSame(newItem!!.composition, currentItem!!.composition)) {
            onCurrentCompositionChanged(newItem, playQueueEvent.trackPosition)
        }
    }

    private fun onCurrentCompositionChanged(newItem: PlayQueueItem?, trackPosition: Long) {
        viewState.showCurrentQueueItem(newItem, isCoversEnabled)
        if (newItem != null
                && (newItem != currentItem || newItem.composition.duration != currentItem!!.composition.duration)) {
            viewState.showTrackState(trackPosition, newItem.composition.duration)
        }
        currentItem = newItem
    }

    private fun subscribeOnPlayerStateChanges() {
        batterySafeDisposable.add(playerInteractor.playerStateObservable
                .observeOn(uiScheduler)
                .subscribe(this::onPlayerStateChanged))
    }

    private fun onPlayerStateChanged(playerState: PlayerState) {
        viewState.showPlayerState(playerState)
    }

    private fun subscribeOnPlayQueue() {
        batterySafeDisposable.add(playerInteractor.playQueueObservable
                .observeOn(uiScheduler)
                .filter(listDragFilter::filterListEmitting)
                .subscribe(this::onPlayQueueChanged, this::onPlayQueueReceivingError))
    }

    private fun onPlayQueueChanged(list: List<PlayQueueItem>) {
        playQueue = list
        viewState.showPlayQueueSubtitle(playQueue.size)
        viewState.setMusicControlsEnabled(playQueue.isNotEmpty())
        viewState.updatePlayQueue(list)
    }

    private fun onPlayQueueReceivingError(throwable: Throwable) {
        errorParser.parseError(throwable)
        viewState.setMusicControlsEnabled(false)
    }

    private fun subscribeOnTrackPositionChanging() {
        batterySafeDisposable.add(playerInteractor.trackPositionObservable
                .observeOn(uiScheduler)
                .subscribe(this::onTrackPositionChanged))
    }

    private fun onTrackPositionChanged(currentPosition: Long) {
        if (currentItem != null) {
            val duration = currentItem!!.composition.duration
            viewState.showTrackState(currentPosition, duration)
        }
    }

    private fun subscribeOnUiSettings() {
        playerScreenInteractor.coversEnabledObservable
                .subscribeOnUi(this::onUiSettingsReceived, errorParser::logError)
    }

    private fun onUiSettingsReceived(isCoversEnabled: Boolean) {
        this.isCoversEnabled = isCoversEnabled
        viewState.setPlayQueueCoversEnabled(isCoversEnabled)
        if (currentItem != null) {
            viewState.showCurrentQueueItem(currentItem, isCoversEnabled)
        }
    }

    private fun onDefaultError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showErrorMessage(errorCommand)
    }

    private fun subscribeOnRandomMode() {
        playerInteractor.randomPlayingObservable
                .unsafeSubscribeOnUi(viewState::showRandomPlayingButton)
    }

    private fun subscribeOnSpeedAvailableState() {
        playerInteractor.speedChangeAvailableObservable
                .unsafeSubscribeOnUi(viewState::showSpeedChangeFeatureVisible)
    }

    private fun subscribeOnSpeedState() {
        playerInteractor.playbackSpeedObservable
                .unsafeSubscribeOnUi(viewState::displayPlaybackSpeed)
    }

    private fun subscribeOnSleepTimerTime() {
        batterySafeDisposable.add(playerScreenInteractor.sleepTimerCountDownObservable
                .observeOn(uiScheduler)
                .subscribe(viewState::showSleepTimerRemainingTime))
    }
}